### PR TITLE
Refactor gcs part 1 random access file

### PR DESCRIPTION
**Pull request by vnghia**
@mihaimaruseac 
This PR refactors `gcs` to use the `core` implementation. The change is:
- Get various envs
- Use `ram_file_block_cache` https://github.com/tensorflow/tensorflow/blob/2629e4f9a4d982448494c880402f603cf559c488/tensorflow/core/platform/cloud/gcs_file_system.cc#L826
- This variable can be controlled by env so I drop it https://github.com/tensorflow/tensorflow/blob/2629e4f9a4d982448494c880402f603cf559c488/tensorflow/core/platform/cloud/gcs_file_system.cc#L788
- Merge 2 `RandomAccessFile` into one. https://github.com/tensorflow/tensorflow/blob/2629e4f9a4d982448494c880402f603cf559c488/tensorflow/core/platform/cloud/gcs_file_system.cc#L261 and https://github.com/tensorflow/tensorflow/blob/2629e4f9a4d982448494c880402f603cf559c488/tensorflow/core/platform/cloud/gcs_file_system.cc#L289
- I am missing this part https://github.com/tensorflow/tensorflow/blob/2629e4f9a4d982448494c880402f603cf559c488/tensorflow/core/platform/cloud/gcs_file_system.cc#L991-L1003 will be added in later PR.

### CODE REVIEW

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc [hunk 1]
```diff
@@ -82,44 +103,106 @@ static void MaybeAppendSlash(std::string* name) {
...
+        // Empty the buffer to avoid caching bad reads.
+        gcs_file->buffer.resize(0);
+        return -1;
+      }
+      size_t remaining_copy =
+          (std::min)(n - copy_size, gcs_file->buffer.size());
+      memcpy(buffer + copy_size, gcs_file->buffer.data(), remaining_copy);
+      copy_size += remaining_copy;
+    }
+    return copy_size;
```
Comment (vnvo2409) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:204]: This part fails on my local test. I will check for it

**Comment by vnvo2409** [comment]
@mihaimaruseac 
Ready to review

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc [hunk 2]
```diff
@@ -82,44 +103,114 @@ static void MaybeAppendSlash(std::string* name) {
...
 
-// TODO(vnvo2409): Adding cache.
+static void FillBuffer(uint64_t start, GCSFile* gcs_file, TF_Status* status) {
+  gcs_file->buffer_start = start;
+  gcs_file->buffer.resize(gcs_file->buffer_size);
+  auto read =
+      gcs_file->read_fn(gcs_file->path, gcs_file->buffer_start,
+                        gcs_file->buffer_size, &(gcs_file->buffer[0]), status);
+  gcs_file->buffer_end_is_past_eof = (TF_GetCode(status) == TF_OUT_OF_RANGE);
+  if (read >= 0) gcs_file->buffer.resize(read);
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:169]:
```
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:164:13: error: writing variable 'buffer_start' requires holding mutex 'gcs_file->buffer_mutex' exclusively [-Werror,-Wthread-safety-analysis]
  gcs_file->buffer_start = start;
            ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:165:13: error: reading variable 'buffer' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
  gcs_file->buffer.resize(gcs_file->buffer_size);
            ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:167:51: error: reading variable 'buffer_start' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
      gcs_file->read_fn(gcs_file->path, gcs_file->buffer_start,
                                                  ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:168:60: error: reading variable 'buffer' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
                        gcs_file->buffer_size, &(gcs_file->buffer[0]), status);
                                                           ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:169:13: error: writing variable 'buffer_end_is_past_eof' requires holding mutex 'gcs_file->buffer_mutex' exclusively [-Werror,-Wthread-safety-analysis]
  gcs_file->buffer_end_is_past_eof = (TF_GetCode(status) == TF_OUT_OF_RANGE);
            ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:170:28: error: reading variable 'buffer' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
  if (read >= 0) gcs_file->buffer.resize(read);
                           ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:468:25: error: reading variable 'file_block_cache' requires holding mutex 'gcs_file->block_cache_lock' [-Werror,-Wthread-safety-analysis]
            ? gcs_file->file_block_cache->Read(path, offset, n, buffer, status)
```

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc [hunk 3]
```diff
@@ -82,44 +103,114 @@ static void MaybeAppendSlash(std::string* name) {
...
-  gcs::Client* gcs_client;  // not owned
+  const std::string path;
+  const bool is_cache_enable;
+  const uint64_t buffer_size;
+  ReadFn read_fn;
+  absl::Mutex buffer_mutex;
+  uint64_t buffer_start ABSL_GUARDED_BY(buffer_mutex);
+  bool buffer_end_is_past_eof ABSL_GUARDED_BY(buffer_mutex);
+  std::string buffer ABSL_GUARDED_BY(buffer_mutex);
+  GCSFile(std::string path, bool is_cache_enable, uint64_t buffer_size,
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:145]:
```suggestion

  GCSFile(std::string path, bool is_cache_enable, uint64_t buffer_size,
```

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc [hunk 4]
```diff
@@ -82,44 +103,114 @@ static void MaybeAppendSlash(std::string* name) {
...
+          ReadFn read_fn)
+      : path(path),
+        is_cache_enable(is_cache_enable),
+        buffer_size(buffer_size),
+        buffer_mutex(),
+        buffer_start(0),
+        buffer_end_is_past_eof(false),
+        buffer(),
+        read_fn(std::move(read_fn)) {}
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:154]:
```suggestion
        is_cache_enable(is_cache_enable),
        buffer_size(buffer_size),
        read_fn(std::move(read_fn)),
        buffer_mutex(),
        buffer_start(0),
        buffer_end_is_past_eof(false),
        buffer() {}
```

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
There are some errors internally:

- [x] 1 linter error fixed by adding a new line
- [x] 1 error due to object construction rules, fixed by reordering the fields
- [x] 1 error due to requiring mutex lock before initializing variables. This one needs more change

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc (hunk 2 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:171]: The errors with `gcs_file->buffer_mutex` are caused by
https://github.com/tensorflow/tensorflow/blob/c4a07915a2868303edf496b32ae675bcbc7a0715/tensorflow/core/platform/cloud/gcs_file_system.cc#L352-L353
I forgot to add `ABSL_EXCLUSIVE_LOCKS_REQUIRED`.

The error with `gcs_file->block_cache_lock` is caused by https://github.com/tensorflow/tensorflow/blob/c4a07915a2868303edf496b32ae675bcbc7a0715/tensorflow/core/platform/cloud/gcs_file_system.cc#L992

Both are fixed now.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
This still fails to build internally

```
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:165:3: error: expected unqualified-id
  gcs_file->buffer_start = start;
```

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc [hunk 5]
```diff
@@ -82,44 +103,116 @@ static void MaybeAppendSlash(std::string* name) {
...
 } GCSFile;
 
 void Cleanup(TF_RandomAccessFile* file) {
   auto gcs_file = static_cast<GCSFile*>(file->plugin_file);
   delete gcs_file;
 }
 
-// TODO(vnvo2409): Adding cache.
+static void FillBuffer(uint64_t start, GCSFile* gcs_file, TF_Status* status) {
+  ABSL_EXCLUSIVE_LOCKS_REQUIRED(gcs_file->buffer_mutex)
```
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:164]:
```suggestion
  ABSL_EXCLUSIVE_LOCKS_REQUIRED(gcs_file->buffer_mutex);
```

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by vnvo2409** [comment]
Apologize for the error. I hope everything is good now.

**Comment by mihaimaruseac** [comment]
Hmm, still fails

```
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:164:3: error: 'exclusive_locks_required' attribute cannot be applied to a statement
  ABSL_EXCLUSIVE_LOCKS_REQUIRED(gcs_file->buffer_mutex);
  ^                                                    ~
absl/base/thread_annotations.h:166:18: note: expanded from macro 'ABSL_EXCLUSIVE_LOCKS_REQUIRED'
  __attribute__((exclusive_locks_required(__VA_ARGS__)))
                 ^
1 error generated.
```

**Comment by vnvo2409** [comment]
`ABSL_EXCLUSIVE_LOCKS_REQUIRED` is usually used for member function of a class which uses some variable member guarded by another member mutex. I dont know how to use it outside of class scope. Maybe we should remove all the `ABSL_GUARDED_BY` ?

**Comment by vnvo2409** [comment]
One more question: Why the status of `import/copybara` is passed ? It is really strange to me.

**Comment by mihaimaruseac** [comment]
For the first question, let me look more into it.

For the copybara question, that check only states that the change has been imported internally. There are no tests being run on `import/copybara` (the internal tests are triggered upon internal approval of the converted code)

**Comment by mihaimaruseac** [comment]
What if we use `absl::MutexLock l(&gcs_file->buffer_mutex);` instead of the absl macro?

**Comment by vnvo2409** [comment]
I got an idea. In the `core` implementation, `FillBuffer` is used only one time. So I think the best solution is to remove `FillBuffer` and just write that code directly into that function.

**Comment by vnvo2409** [comment]
> What if we use `absl::MutexLock l(&gcs_file->buffer_mutex);` instead of the absl macro?

`Read` is holding `absl::MutexLock l(&gcs_file->buffer_mutex);`, so it may cause error if we try to lock `gcs_file->buffer_mutex` again.

**Comment by mihaimaruseac** [comment]
Good point. We can make it an internal function of the struct or just inline. Let's go with inlining for now since it has only one use

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**